File: backend/src/modules/v1-pilot/alpha/meta-alpha.combiner.ts
import { AlphaDirection } from '../contracts/v1-pilot.contracts';

export const META_NUMERIC_WEIGHT = 0.5;
export const META_EVENT_WEIGHT = 0.25;
export const META_MACRO_WEIGHT = 0.15;
export const META_RISK_WEIGHT = 0.1;
export const META_UP_THRESHOLD = 0.65;

export interface MetaDirectionInput {
  direction?: AlphaDirection;
  confidence?: number;
}

export interface MetaComponentScores {
  numericScore: number;
  eventScore: number;
  macroScore: number;
  riskAdjustment: number;
  finalScore: number;
}

export interface MetaDecisionExportRecord {
  id: string;
  symbol: string;
  asOf: string;
  availableAt: string;
  horizonHours: number;
  direction: 'up' | 'flat';
  confidence: number;
  numericScore: number;
  eventScore: number;
  macroScore: number;
  riskAdjustment: number;
  finalScore: number;
  llmScores: {
    event: number;
    macro: number;
    riskAdjustment: number;
  };
  maxPositionPct?: number;
  featureSnapshotHash: string;
  evidenceRefs: string[];
  llmFeatureRefs: string[];
  numericFeatureRefs: string[];
  outputHash: string;
}

export function directionScore(
  direction: AlphaDirection | undefined,
  confidence = 0.5,
): number {
  if (direction === 'up') {
    return confidence;
  }
  if (direction === 'down') {
    return 1 - confidence;
  }
  return 0.5;
}

export function combineMetaComponentScores(
  numericScore: number,
  eventScore: number,
  macroScore: number,
  riskAdjustment: number,
): number {
  const finalScore =
    numericScore * META_NUMERIC_WEIGHT +
    eventScore * META_EVENT_WEIGHT +
    macroScore * META_MACRO_WEIGHT +
    riskAdjustment * META_RISK_WEIGHT;
  return Number(Math.min(1, Math.max(0, finalScore)).toFixed(6));
}

export function combineMetaFromDecisions(input: {
  numeric?: MetaDirectionInput;
  llmEvent?: MetaDirectionInput;
  llmMacro?: MetaDirectionInput;
  llmRisk?: MetaDirectionInput;
}): MetaComponentScores {
  const numericScore = directionScore(
    input.numeric?.direction,
    input.numeric?.confidence,
  );
  const eventScore = directionScore(
    input.llmEvent?.direction,
    input.llmEvent?.confidence,
  );
  const macroScore = directionScore(
    input.llmMacro?.direction,
    input.llmMacro?.confidence,
  );
  const riskAdjustment =
    1 - directionScore(input.llmRisk?.direction, input.llmRisk?.confidence);
  const finalScore = combineMetaComponentScores(
    numericScore,
    eventScore,
    macroScore,
    riskAdjustment,
  );
  return { numericScore, eventScore, macroScore, riskAdjustment, finalScore };
}

export function directionFromMetaScore(finalScore: number): 'up' | 'flat' {
  return finalScore >= META_UP_THRESHOLD ? 'up' : 'flat';
}

/** Mirrors LEAN replay path: prefer Nest-exported component scores and finalScore. */
export function combineMetaLeanReplay(
  liveNumericScore: number,
  metaRecord: Partial<MetaDecisionExportRecord> | null | undefined,
): MetaComponentScores & { direction: 'up' | 'flat' } {
  if (!metaRecord) {
    return {
      numericScore: liveNumericScore,
      eventScore: 0.5,
      macroScore: 0.5,
      riskAdjustment: 0.5,
      finalScore: liveNumericScore,
      direction: directionFromMetaScore(liveNumericScore),
    };
  }

  const llmScores = metaRecord.llmScores;
  const numericScore = metaRecord.numericScore ?? liveNumericScore;
  const eventScore = metaRecord.eventScore ?? llmScores?.event ?? 0.5;
  const macroScore = metaRecord.macroScore ?? llmScores?.macro ?? 0.5;
  const riskAdjustment =
    metaRecord.riskAdjustment ?? llmScores?.riskAdjustment ?? 0.5;
  const finalScore =
    metaRecord.finalScore ??
    combineMetaComponentScores(
      numericScore,
      eventScore,
      macroScore,
      riskAdjustment,
    );

  return {
    numericScore,
    eventScore,
    macroScore,
    riskAdjustment,
    finalScore,
    direction: directionFromMetaScore(finalScore),
  };
}

File: engines/lean/aggressive_llm_momentum/alpha/meta_alpha_combiner.py
"""Pure meta-alpha combiner shared by LEAN replay and parity tests."""

from __future__ import annotations

NUMERIC_WEIGHT = 0.50
EVENT_WEIGHT = 0.25
MACRO_WEIGHT = 0.15
RISK_WEIGHT = 0.10
UP_THRESHOLD = 0.65


def direction_score(direction: str | None, confidence: float = 0.5) -> float:
    if direction == "up":
        return confidence
    if direction == "down":
        return 1.0 - confidence
    return 0.5


def combine_meta_component_scores(
    numeric_score: float,
    event_score: float,
    macro_score: float,
    risk_adjustment: float,
) -> float:
    final_score = (
        NUMERIC_WEIGHT * numeric_score
        + EVENT_WEIGHT * event_score
        + MACRO_WEIGHT * macro_score
        + RISK_WEIGHT * risk_adjustment
    )
    return round(max(0.0, min(1.0, final_score)), 6)


def direction_from_meta_score(final_score: float) -> str:
    return "up" if final_score >= UP_THRESHOLD else "flat"


def resolve_component_scores(
    live_numeric_score: float,
    meta_record: dict | None,
) -> dict[str, float]:
    if meta_record is None:
        return {
            "numericScore": live_numeric_score,
            "eventScore": 0.5,
            "macroScore": 0.5,
            "riskAdjustment": 0.5,
            "finalScore": live_numeric_score,
        }

    llm_scores = meta_record.get("llmScores") or {}
    numeric_score = float(meta_record.get("numericScore", live_numeric_score))
    event_score = float(meta_record.get("eventScore", llm_scores.get("event", 0.5)))
    macro_score = float(meta_record.get("macroScore", llm_scores.get("macro", 0.5)))
    risk_adjustment = float(
        meta_record.get("riskAdjustment", llm_scores.get("riskAdjustment", 0.5)),
    )
    if meta_record.get("finalScore") is not None:
        final_score = float(meta_record["finalScore"])
    else:
        final_score = combine_meta_component_scores(
            numeric_score,
            event_score,
            macro_score,
            risk_adjustment,
        )

    return {
        "numericScore": numeric_score,
        "eventScore": event_score,
        "macroScore": macro_score,
        "riskAdjustment": risk_adjustment,
        "finalScore": final_score,
    }
